text/000-dynamic-components.md

```diff
@@ -0,0 +1,338 @@
+
+## Guiding Principles
+
+We want to avoid ad-hoc solutions and proprietary non-standard ways to define dynamic component loading. We want to build an API that is future proof, whenever more things land into the spec for modules.
+
+## General Invariants
+
+Note that the invariants defined bellow are not necessarily constraints on LWC in general but rather explicitly defined for the UI Platform use cases:
+
+- A dynamic import must always be statically verifiable.
+    - A dependency should never be discovered or learned of for the first time at runtime.
+    - Note that the ability to analyze is orthogonal to when a component is loaded.
+- It must not break the reactivity model or the general framework lifecycle.
+    - Once the component is loaded it should function within the lifecycle of the parent component.
+    - No special handling in user-land is needed.
+- It must work regardless of the algorithm, even if nothing is preloaded at all.
+    - The decision for when a dynamic dependency is loaded is orthogonal to its definition.
+    - We might have a preload mechanism based on MRU or on a specific context (ex.mobile) but the component should work regardless of the algorithm even if nothing is preloaded at all.
+
+You will notice that this follows the same pattern as what we have done for `@wire`: Bring the ability to understand and resolve dependencies at compile, build and design time.
+
+## API
+
+Bellow is the API and its ergonomics for dynamically loading a module. It has very similar mental model to the `is` attribute proposal (which we can't use since Safari will not implement it in the foreseeable future): An arbitrary tag name on which you can assign any constructor.
+
+```json5
+{
+    "dynamicComponent": {
+        "loader": "@salesforce/loader",
+        "strict": true // only allow string identifiers
+    }
+}
+```
+
+```html
+<!-- original HTML source -->
+<template>
+  <lazy-component lwc:dynamic={lazyConstructor} value={value} other="test"></lazy-component>
+</template>
+
+
+<!-- original JS source -->
+<script>
+import { LightningElement, track } from "lwc";
+export default class LazyTest extends LightningElement {
+    @api value = "yay!";
+    @track lazyConstructor;
+
+    async connectedCallback() {
+      const { default: Ctor } = await import("lightning/concreteComponent");
+      this.lazyConstructor = Ctor;
+    }
+}
+</script>
+
+<!-- compiled code -->
+<script>
+  import { loader } from "@salesforce/loader";
+  // this can't be imported in user-land
+
+  function tmpl($api, $cmp) {
+     return [$cmp.lazyConstructor != null // falsy for null or undefined
```

Comment:
```suggestion
     return [$cmp.lazyConstructor != null ? // falsy for null or undefined
```

Reply:
it looks like this is supposed to be a ternary but is missing the "`?`".

Reply:
yeah that is psudo-code, will fix